### PR TITLE
Lazily load samples for unit testing

### PR DESCRIPTION
**Pull request by codesome**
Closes https://github.com/prometheus/prometheus/issues/4838

Scenario:
* After initially adding all samples to test suite, the `minValidTime` of the head is incremented at times.
* So when we add samples for recording rules, they are rejected, as they are from the start.
* Hence lazily adding samples to the test suite instead of adding all at the beginning.

### CODE REVIEW

**Comment by gouthamve** [comment]
Hmm, while I love this approach, this is to me points to the fact that we are also writing stuff down to disk as part of the tests. I am not super happy about that and think we can change the behavior by having super large block size.

**Comment by codesome** [comment]
What would be a good value for the min block duration?

**Comment by gouthamve** [comment]
I guess, we will be writing to disk for now. I'll review the lazy loading now.

File: cmd/promtool/unittest.go [hunk 1]
```diff
@@ -191,8 +186,17 @@ func (tg *testGroup) test(mint, maxt time.Time, evalInterval time.Duration, grou
...
+		suite.WithSamplesTill(ts, func(err error) {
+			if err != nil {
+				errs = append(errs, err)
+				return
+			}
+			for _, g := range groups {
+				g.Eval(suite.Context(), ts)
+			}
+		})
+		if len(errs) > 0 {
```
[the remark this answers is not in the document]
Comment (codesome) [reply cmd/promtool/unittest.go]: We need this to catch any error in `suite.WithSamplesTill` call right?

[the remark this answers is not in the document]
Comment (gouthamve) [reply cmd/promtool/unittest.go]: Oh, I see now. Didn't know what I was thinking then.

**Comment by codesome** [comment]
@gouthamve is this good to be merged?

**Pull request merged**